prep/go/tutorial/chapter4/xkcd.go

```diff
@@ -0,0 +1,465 @@
+package main
+
+import (
+        "os"
+        "fmt"
+        "strconv"
+        "strings"
+        "log"
+        "encoding/json"
+        "net/http"
+        "io/ioutil"
+        "sync"
+        "unicode"
+        "bufio"
+        "sort"
+)
+
+type XKCDComic struct {
+    Month string `json:"month"`
+    Id int `json:"num"`
+    Link string `json:"link"`
+    Year string `json:"year"`
+    News string `json:"new"`
+    SafeTitle string `json:"safe_title"`
+    Transcript string `json:"transcript"`
+    Alt string `json:"alt"`
+    Img string `json:"img"`
+    Title string `json:"title"`
+    Day string `json:"day"`
+    Indexed bool 
+}
+
+type XKCDComicView struct {
+    Id int `json:"id"`
+    Title string `json:"title"`
+    TranscriptSummary string `json:"transcript_summary"`
+    Link string `json:"link"`
+}
+
+type ComicIndex map[string]map[int]bool
+
+func newComicIndex() ComicIndex {
+    var index ComicIndex
+    index = make(map[string]map[int]bool)
+    return index
+}
+
+func addComicIndex(index ComicIndex, term string, id int) {
+    if index[term] == nil {
+        index[term] = make(map[int]bool)
+    }
+    index[term][id] = true
+}
+
+
+func toView(comic XKCDComic) XKCDComicView {
+    var view XKCDComicView
+    view.Id = comic.Id
+    view.Title = comic.Title
+    // Some comics don't have transcripts
+    if len(comic.Transcript) >= 100 {
+        view.TranscriptSummary = comic.Transcript[:100]
+    } else {
+        view.TranscriptSummary = comic.Transcript
+    }
+    view.TranscriptSummary += "..."
+    view.Link = fmt.Sprintf(XKCD_URL_FMT, comic.Id)
+    return view
+}
+
+var XKCD_URL_JSON_FMT string 
+var XKCD_URL_FMT string 
+var COMIC_STORAGE_FMT string 
+var COMIC_INDEX_STORAGE string
+
+
+func mkComicStorageDir() string { 
+    comicStorageDir := "/.local/xkcd_cmd/"
+    homeDir, err := os.UserHomeDir()
+    if err != nil {
+        log.Fatal(err)
+        os.Exit(1)
```

Comment:
- If you have `log.Fatal` you don't need a subsequent `os.Exit`, since the `log.Fatal` will crash the program already
- One stylistic approach that might make life easier (at the cost of making code a bit more verbose) is to have your functions return the error (e.g. `(string, error)` in this case), and push the `log.Fatal` or `os.Exit` calls as high as possible (e.g. in `main`); that way, you don't have to always think about "could calling this function cause the program to crash"